### PR TITLE
Bug: Mapper uses tf data for drone pose, which can be delayed while communicating with the basestation

### PR DESCRIPTION
**Pull request by marcelino-pensa**
**Related Issue**
closes https://github.com/Pensasystems/pensa-ros/issues/1016

**Describe the solution**
- Get rid of tf threads and use synchronized subscribers for lidar and pose

**How has this been tested**

TEST CARD: https://docs.google.com/spreadsheets/d/1JSrwVr8BYjKtopL1Z9jKnUWDNz2CmTmCUixyLyMLfOE/edit#gid=486985852

**Type of Changes**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New node addition
- [ ] Introduces a new third-party library 
  
**Checklist**

- [x] Created draft PR for review before starting formal testing
- [ ] Addressed comments in draft PR
- [ ] Completed test card and filled in all relevant information

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -10,6 +10,7 @@ add_compile_options(-std=c++11)
 find_package(catkin REQUIRED COMPONENTS
   cmake_modules
   geometry_msgs
+  message_filters
```
Comment (marcelino-pensa) [inline CMakeLists.txt:13]: for having synchronized topic subscriptions

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,6 +19,8 @@ find_package(catkin REQUIRED COMPONENTS
   sensor_msgs
   std_msgs
   std_srvs
+  tf2
+  tf2_ros
```
Comment (marcelino-pensa) [inline CMakeLists.txt:23]: for some convenient functions within the `tf2` library

File: include/mapper/helper.h [hunk 3]
```diff
@@ -52,4 +56,12 @@ struct timespec TimeFromNow(const uint& increment_sec);
 // Function that returns whether value1 and value2 are close enough (for double inputs)
 bool AreDoubleApproxEqual(const double &value1, const double &value2, const double &epsilon);
 
+// Convert geometry_msgs::Pose to tf2::Transform
+tf2::Transform PoseToTransform(const geometry_msgs::Pose &pose);
+
+// Function to lookup an arbitrary transform in the tf tree
+bool LookupTransform(const std::string &from_frame,
+                     const std::string &to_frame,
+                     tf2::Transform *tf_output);
```
Comment (marcelino-pensa) [inline include/mapper/helper.h]: We use this to find out the transform between the drone's base_link and lidar. We call this once when starting up `mapper`

File: include/mapper/mapper_class.h [hunk 4]
```diff
@@ -27,13 +27,15 @@
...
+#include <message_filters/subscriber.h>
+#include <message_filters/sync_policies/approximate_time.h>
+#include <message_filters/time_synchronizer.h>
 #include <ros/ros.h>
 #include <ros/publisher.h>
 #include <ros/subscriber.h>
 #include <std_msgs/Float32.h>
 #include <std_srvs/Trigger.h>
 #include <std_srvs/SetBool.h>
-#include <tf/transform_listener.h>
```
Comment (marcelino-pensa) [inline include/mapper/mapper_class.h:36]: We used to listen to Lidar's pose in `store` frame by listening to the transform tree, which we will no longer do

File: include/mapper/msg_conversions.h [hunk 5]
```diff
@@ -45,10 +45,8 @@ namespace msg_conversions {
   void                   ros_to_array_point(const geometry_msgs::Point & p, float* array);
   geometry_msgs::Point   array_to_ros_point(float* array);
   geometry_msgs::Point   set_ros_point(const double & x, const double & y, const double & z);
-  geometry_msgs::Point   pcl_to_ros_vector(const pcl::PointXYZ& pt);
```
Comment (marcelino-pensa) [inline include/mapper/msg_conversions.h:48]: Rename function

File: include/mapper/msg_conversions.h [hunk 6]
```diff
@@ -45,10 +45,8 @@ namespace msg_conversions {
...
-
-  Eigen::Vector3d        tf_vector3_to_eigen_vector(const tf::Vector3 & v);
```
Comment (marcelino-pensa) [inline include/mapper/msg_conversions.h:50]: unused function in code

File: include/mapper/octoclass.h [hunk 7]
```diff
@@ -87,11 +86,11 @@ class OctoClass{
...
                                     sensor_msgs::PointCloud2& cloud);  // Convert from octomap to pointcloud2
     void PclToRayOctomap(const pcl::PointCloud< pcl::PointXYZ > &cloud,
-                          const tf::StampedTransform &tf_cam2world,
-                          const algebra_3d::FrustumPlanes &frustum);    // Map obstacles and free area
+                         const tf2::Transform &tf_pcl2world,
+                         const algebra_3d::FrustumPlanes &frustum);    // Map obstacles and free area
     // Same as above, but without frustum filtering (for lidar)
     void PclToRayOctomap(const pcl::PointCloud< pcl::PointXYZ > &cloud,
-                          const tf::StampedTransform &tf_cam2world);
+                          const tf2::Transform &tf_pcl2world);
```
Comment (marcelino-pensa) [inline include/mapper/octoclass.h:93]: Updated the inputs to use `tf2` instead of `tf`, just for convenience

File: include/mapper/structs.h [hunk 8]
```diff
@@ -39,21 +39,21 @@
 
 namespace mapper {
 
-struct stampedPcl {
+struct StampedPcl {
     pcl::PointCloud<pcl::PointXYZ> cloud;
-    tf::StampedTransform tf_cam2world;
+    tf2::Transform tf_pcl2world;
```
Comment (marcelino-pensa) [inline include/mapper/structs.h:44]: `pcl::PointCloud` is already stamped, so we don't need two stamped variables in here

File: include/mapper/structs.h [hunk 9]
```diff
@@ -39,21 +39,21 @@
...
     bool is_lidar;
 };
 
 struct globalVariables {
     // Mutex protected variables
-    tf::StampedTransform tf_body2world;
+    tf2::Transform tf_body2world;
```
Comment (marcelino-pensa) [inline include/mapper/structs.h:50]: We were not using the `stamp` in this variable for anything, so I removed the stamp from it (beyong moving from `tf` to `tf2`)

File: include/mapper/structs.h [hunk 10]
```diff
@@ -39,21 +39,21 @@
 
 namespace mapper {
 
-struct stampedPcl {
+struct StampedPcl {
```
Comment (marcelino-pensa) [inline include/mapper/structs.h:42]: Rename to conform with the style guide

File: launch/mapper.launch [hunk 11]
```diff
@@ -5,13 +5,10 @@
     <arg name="process_pcl_at_startup" default="true" />
     <arg name="map_3d" default="true" />
     <arg name="inertial_frame_id" default="/map" />
-    <arg name="robot_frame_id" default="/robot" />
-    <arg name="depth_cam_prefix" default="/" />
-    <arg name="depth_cam_suffix" default="/camera/depth/compressed_points" />
-    <arg name="lidar_names" default="['/laser/scan/cloud']" />
-    <arg name="lidar_prefix" default="/" />
-    <arg name="lidar_frame_id" default="['/laser_frame']" />
-    <arg name="lidar_suffix" default="" />
```
Comment (marcelino-pensa) [inline launch/mapper.launch:14]: We used to have a system that would allow to have multiple depth cameras and lidar sensors. I am getting rid of this whole mechanism, but I might remake this in the future if we ever use `mapper` to subscribe to multiple PCL topics

File: src/callbacks.cpp [hunk 12]
```diff
@@ -23,48 +23,19 @@
...
-                                    const uint& cam_index) {
-    // Structure to include pcl and its frame
-    stampedPcl new_pcl;
-    const uint max_queue_size = globals_.max_queue_size;
-
-    // Convert message into pcl type
-    pcl::PointCloud< pcl::PointXYZ > cloud;
-    pcl_conversions::FromROSMsg(*msg, &cloud);
-    new_pcl.cloud = cloud;
+void MapperClass::LidarSyncCallback(const sensor_msgs::PointCloud2::ConstPtr &lidar_msg,
```
Comment (marcelino-pensa) [inline src/callbacks.cpp:26]: Get lidar scan and base_link pose, add both to a queue for being processed by the octomap thread

File: src/callbacks.cpp [hunk 13]
```diff
@@ -79,6 +50,14 @@ void MapperClass::LidarPclCallback(const sensor_msgs::PointCloud2::ConstPtr &msg
     mutexes_.point_cloud.unlock();
 }
 
+void MapperClass::BaseLinkPoseCallback(const geometry_msgs::PoseStamped::ConstPtr &msg) {
```
Comment (marcelino-pensa) [inline src/callbacks.cpp:54]: Callback that just subscribes to the robot's base_link pose, and update a global variable with the latest pose

File: src/callbacks.cpp [hunk 14]
```diff
@@ -223,15 +202,11 @@ void MapperClass::TrajectoryStatusCallback(const pensa_msgs::trapezoidal_p2pActi
...
     waypoints_sub_.shutdown();
-    for (uint i = 0; i < cameras_sub_.size(); i++) {
-        cameras_sub_[i].shutdown();
-    }
-    for (uint i = 0; i < lidar_sub_.size(); i++) {
-        lidar_sub_[i].shutdown();
-    }
+    base_link_pose_sub_.shutdown();
+    lidar_sync_sub_->unsubscribe();
+    base_link_pose_sync_sub_->unsubscribe();
```
Comment (marcelino-pensa) [inline src/callbacks.cpp:210]: I don't really remember why I had to destroy all callbacks, but I've updated this function anyway

File: src/threads.cpp [hunk 15]
```diff
@@ -51,72 +51,6 @@ void MapperClass::FadeTask() {
     ROS_DEBUG("[mapper]: Exiting Fading Memory Thread...");
 }
 
-// Thread for updating the body tfTree values
```
Comment (marcelino-pensa) [inline src/threads.cpp:54]: deleted threads that would be listening to the tf tree

File: src/threads.cpp [hunk 16]
```diff
@@ -297,22 +231,17 @@ void MapperClass::OctomappingTask() {
         // Get Point Cloud
         mutexes_.point_cloud.lock();
             // Get data from queue
-            pcl::PointCloud< pcl::PointXYZ > point_cloud = globals_.pcl_queue.front().cloud;
```
Comment (marcelino-pensa) [inline src/threads.cpp:300]: rename variable for clarity

File: src/threads.cpp [hunk 17]
```diff
@@ -297,30 +231,19 @@ void MapperClass::OctomappingTask() {
...
             const bool is_lidar = globals_.pcl_queue.front().is_lidar;
 
             // Remove data from queue
             globals_.pcl_queue.pop();
         mutexes_.point_cloud.unlock();
 
-        // Check if a tf message has been received already. If not, return
-        if (tf_cam2world.stamp_.toSec() == 0) {
-            continue;
-        }
```
Comment (marcelino-pensa) [inline src/threads.cpp:311]: No need to check this anymore

File: src/threads.cpp [hunk 18]
```diff
@@ -297,30 +231,19 @@ void MapperClass::OctomappingTask() {
...
-
-        // Get camera transform
-        tf::Quaternion q = tf_cam2world.getRotation();
-        tf::Vector3 v = tf_cam2world.getOrigin();
-        Eigen::Affine3d transform = Eigen::Affine3d::Identity();
-        transform.translation() << v.getX(), v.getY(), v.getZ();
-        transform.rotate(Eigen::Quaterniond(q.getW(), q.getX(), q.getY(), q.getZ()));
```
Comment (marcelino-pensa) [inline src/threads.cpp:318]: This portion of the code was sent to `msg_conversions::tf_transform_to_eigen_transform`

File: src/threads.cpp [hunk 19]
```diff
@@ -297,30 +231,19 @@ void MapperClass::OctomappingTask() {
...
-
         // Update frustum orientation
+        const Eigen::Affine3d eigen_pcl2world = msg_conversions::tf_transform_to_eigen_transform(tf_pcl2world);
```
Comment (marcelino-pensa) [inline src/threads.cpp:243]: renamed variable `transform` into `eigen_pcl2world` for better clarity

**Review by marcelino-pensa — COMMENTED** [review]
added a few personal comments

File: launch/mapper_pensa.launch [hunk 20]
```diff
@@ -1,20 +1,20 @@
 <launch>
```
Comment (marcelino-pensa) [inline launch/mapper_pensa.launch:1]: Note: there will be an accompanying PR in `pensa-ros` to accomodate with updated variable names in the launch file

File: include/mapper/helper.h [hunk 21]
```diff
@@ -9,7 +9,11 @@
 #include <geometry_msgs/Point.h>
 #include <octomap/octomap.h>
 #include <pcl/point_cloud.h>
+#include <tf2/LinearMath/Transform.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
```
Comment (adam-pensa) [inline include/mapper/helper.h]: Only one of these should be needed - whichever defines `tf2::Transform`

Comment (adam-pensa) [reply include/mapper/helper.h]: The other can move to the `.cpp`

File: include/mapper/mapper_class.h [hunk 22]
```diff
@@ -219,15 +213,17 @@ class MapperClass {
...
   // Subscriber variables
-  ros::Subscriber trajectory_sub_, trajectory_status_sub_;
+  ros::Subscriber base_link_pose_sub_;
+  ros::Subscriber trajectory_status_sub_;
   ros::Subscriber waypoints_sub_;
-  std::vector<ros::Subscriber> cameras_sub_;
-  std::vector<ros::Subscriber> lidar_sub_;
+
+  // Synchronized subscribers
+  boost::shared_ptr<Sync> sync_;
```
Comment (adam-pensa) [inline include/mapper/mapper_class.h:224]: That's one heavily-nested `typedef`! (nothing wrong with it, just surprised)

File: include/mapper/octoclass.h (hunk 7 above)
Comment (adam-pensa) [reply include/mapper/octoclass.h:93]: Why did the `Stamped` go away?

File: launch/mapper_pensa.launch [hunk 23]
```diff
@@ -1,20 +1,20 @@
...
 
-  <arg name="robot_frame_id"    default="/c0001_bl" />
-  <arg name="inertial_frame_id" default="store" />
+  <arg name="robot_base_link_frame_id" default="/c0001_bl" />
```
Comment (adam-pensa) [inline launch/mapper_pensa.launch]: I suggest removing the default since it's clearly not useful any more

File: launch/mapper_pensa.launch [hunk 24]
```diff
@@ -1,20 +1,20 @@
...
+  <arg name="inertial_frame_id"        default="lp0" />
```
Comment (adam-pensa) [inline launch/mapper_pensa.launch]: same here - if we can start getting rid of `lp0`s now, it will help us transition to multi-perch in the future

File: launch/mapper_pensa.launch [hunk 25]
```diff
@@ -1,20 +1,20 @@
...
-    <arg name="process_pcl_at_startup" value="false" />
-    <arg name="map_3d"                 value="false" />
-    <arg name="inertial_frame_id"      value="$(arg inertial_frame_id)" />
-    <arg name="lidar_frame_id"         value="['/laser_frame']" />
-    <arg name="lidar_names"            value="['/laser/cloud']" />
-    <arg name="robot_frame_id"         value="$(arg robot_frame_id)" />
-    <arg name="max_range"              value="6.0" />
-    <arg name="min_range"              value="0.5" />
-    <arg name="memory_time"            value="-1.0" />
+    <arg name="namespace"                value="cinderella" />
```
Comment (adam-pensa) [inline launch/mapper_pensa.launch]: and probably this one too...

File: src/helper.cpp [hunk 26]
```diff
@@ -98,4 +102,22 @@ bool AreDoubleApproxEqual(const double &value1, const double &value2, const doub
...
+    // Create tf listener
+    static tf2_ros::Buffer tf_buffer;
+    static tf2_ros::TransformListener tf_listener(tf_buffer);
+    try {
+        const geometry_msgs::TransformStamped transform =
+            tf_buffer.lookupTransform(from_frame, to_frame, ros::Time(0), ros::Duration(10.0));
+        tf2::fromMsg(transform.transform, *tf_output);
+    } catch (tf2::TransformException &ex) {
+        ROS_WARN("[mapper]: Transform between frame \"%s\" and frame \"%s\" was not found: %s",
+                 from_frame.c_str(), to_frame.c_str(), ex.what());
```
Comment (adam-pensa) [inline src/helper.cpp]: Why a warning, and not an error?

File: src/helper.cpp [hunk 27]
```diff
@@ -98,4 +102,22 @@ bool AreDoubleApproxEqual(const double &value1, const double &value2, const doub
   return std::fabs(value1 - value2) < epsilon;
 }
 
+bool LookupTransform(const std::string &from_frame,
+                     const std::string &to_frame,
+                     tf2::Transform *tf_output) {
```
Comment (adam-pensa) [inline src/helper.cpp:108]: not sure how I feel about this function. The first time it's called, it won't have any memory of tf topics that were published before the call, since the static variables aren't initialized until first call. So I could see the first call being very long or missing previously-published tfs.

File: src/msg_conversions.cpp [hunk 28]
```diff
@@ -106,16 +106,19 @@ geometry_msgs::Point set_ros_point(const double & x, const double & y, const dou
...
+geometry_msgs::Point tf_vector3_to_ros_point(const tf2::Vector3 & v) {
+  return set_ros_point(v.getX(), v.getY(), v.getZ());
 }
 
-geometry_msgs::Point tf_vector3_to_ros_point(const tf::Vector3 & v) {
-  return set_ros_point(v.getX(), v.getY(), v.getZ());
+tf2::Transform ros_pose_to_tf2_transform(const geometry_msgs::Pose &pose) {
+  tf2::Vector3 pos(pose.position.x, pose.position.y, pose.position.z);
+  tf2::Quaternion q(pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w);
+  tf2::Transform tf2_transform(q, pos);
```
Comment (adam-pensa) [inline src/msg_conversions.cpp]: Any reason you don't do this in one line? I think the compiler will optimize it either way, but just curious.

File: src/threads.cpp (hunk 16 above)
Comment (adam-pensa) [reply src/threads.cpp:300]: The name `pcl_sensor_frame` sounds like it's a frame, not a point cloud. I would suggest something like `cloud_in_sensor_frame` or `sensor_frame_cloud`

File: src/mapper_class.cpp [hunk 29]
```diff
@@ -258,40 +235,30 @@ void MapperClass::Initialize(ros::NodeHandle *nh) {
...
-        std::string lidar_topic = lidar_prefix + lidar_names[i] + lidar_suffix;
-        lidar_sub_[i] = nh->subscribe<sensor_msgs::PointCloud2>
-              (lidar_topic, 10, boost::bind(&MapperClass::LidarPclCallback, this, _1, i));
-        h_lidar_tf_thread_[i] = std::thread(&MapperClass::LidarTfTask, this, inertial_frame_id_, lidar_frame_id[i], i);
-        ROS_INFO("[mapper] Subscribed to lidar topic: %s", lidar_sub_[i].getTopic().c_str());
-    }
+    // Lidar synchronized subscriber
+    lidar_sync_sub_ = new message_filters::Subscriber<sensor_msgs::PointCloud2>(*nh, lidar_topic, 10);
+    base_link_pose_sync_sub_ =
+        new message_filters::Subscriber<geometry_msgs::PoseStamped>(*nh, base_link_pose_topic, 10);
```
Comment (adam-pensa) [inline src/mapper_class.cpp:248]: Do we want to use a queue for these topics? We may want to always use the very latest data for syncing. Can the sync policy have a queue, even if these subs don't? Not quite sure how these two queue sizes interact.

File: src/mapper_class.cpp [hunk 30]
```diff
@@ -80,34 +79,20 @@ void MapperClass::Initialize(ros::NodeHandle *nh) {
...
 
-    // Check if number of cameras/lidar added match the number of frame_id for each of them
-    if (depth_cam_names.size() != cam_frame_id.size()) {
-        ROS_ERROR("[mapper]: Number of cameras is different from camera tf frame_ids!");
-    }
-    if (lidar_names.size() != lidar_frame_id.size()) {
-        ROS_ERROR("[mapper]: Number of lidar topics is different from lidar tf frame_ids!");
+    // Load transform between base_link and lidar
+    if (!helper::LookupTransform(robot_base_link_frame_id, lidar_frame_id, &tf_lidar_in_base_link_frame_)) {
+        return;
```
Comment (adam-pensa) [inline src/mapper_class.cpp:92]: If there's not an error log here, I would definitely make the LookupTransform method throw an error, since this is a critical lookup

File: src/callbacks.cpp (hunk 14 above)
Comment (adam-pensa) [reply src/callbacks.cpp:210]: yeah, I think this happens automatically in subscriber destructors

**Review by adam-pensa — COMMENTED** [review]
Some minor notes, but I think I understand what you're doing and it looks pretty straightforward.

File: include/mapper/octoclass.h (hunk 7 above)
Comment (marcelino-pensa) [reply include/mapper/octoclass.h:93]: We had no need to be passing around `StampedTransform`, as we were not using the stamps anywhere.

File: src/helper.cpp (hunk 26 above)
Comment (marcelino-pensa) [reply src/helper.cpp]: I will switch this to an error

File: src/msg_conversions.cpp (hunk 28 above)
Comment (marcelino-pensa) [reply src/msg_conversions.cpp]: I did it most for readability, as if I got it all in one line it would go over the 120 character limit of `roslint`... It would need at least 3 lines to get it all together, so there wasn't much to gain

File: src/mapper_class.cpp (hunk 29 above)
Comment (marcelino-pensa) [reply src/mapper_class.cpp:248]: I didn't pay much attention to this when I wrote the code, but after reading the [documentation](http://wiki.ros.org/message_filters/ApproximateTime) it seems like it would suffice to have the both with a `queue` of size 2 to best synchronize them... do you also understand it this way? A queue of 10 would just be slightly less efficient, but it wouldn't make a difference

File: launch/mapper.launch [hunk 31]
```diff
@@ -9,7 +8,7 @@
     <arg name="lidar_topic" default="/laser/scan/cloud" />
     <arg name="base_link_pose_topic" default="/vision_pose/pose_base_link" />
     <arg name="lidar_frame_id" default="laser_frame" />
-    <arg name="map_resolution" default="0.1" />
+    <arg name="map_resolution" default="0.075" />
```
Comment (adam-pensa) [inline launch/mapper.launch:11]: Why are we decreasing the map resolution? Won't this make our bags even bigger?

File: launch/mapper.launch [hunk 32]
```diff
@@ -50,8 +46,8 @@
         <param name="fading_memory_update_rate" value="1"/>    <!-- Hz -->
 
         <!-- Map inflation parameters -->
-        <param name="inflate_radius_xy" value="0.30"/>         <!-- meters -->
-        <param name="inflate_radius_z" value="0.30"/>          <!-- meters -->
+        <param name="inflate_radius_xy" value="0.225"/>       <!-- meters -->
+        <param name="inflate_radius_z" value="0.225"/>        <!-- meters -->
```
Comment (adam-pensa) [inline launch/mapper.launch]: And what's the motivation for this change?

File: src/threads.cpp [hunk 33]
```diff
@@ -308,7 +308,7 @@ void MapperClass::OctomappingTask() {
         if ((!is_lidar) && (cam_frustum_pub_.getNumSubscribers() > 0)) {
             visualization_msgs::Marker frustum_markers;
             mutexes_.octomap.lock();
-                globals_.octomap.cam_frustum_.VisualizeFrustum(pcl_sensor_frame.header.frame_id, &frustum_markers);
+               globals_.octomap.cam_frustum_.VisualizeFrustum(cloud_in_sensor_frame.header.frame_id, &frustum_markers);
```
Comment (adam-pensa) [inline src/threads.cpp]: I think you still want the leading space, otherwise it's indented 3 spaces

**Comment by adam-pensa** [comment]
looks good to me, also tested in simulation with our stack

**Review by adam-pensa — APPROVED** [review]

File: launch/mapper.launch (hunk 32 above)
Comment (marcelino-pensa) [reply launch/mapper.launch]: This is not a change in this PR, it was a change in https://github.com/Pensasystems/mapper/pull/5
